disco_aws_automation/disco_datapipeline.py

```diff
@@ -0,0 +1,300 @@
+        """Create a new AsiaqDataPipeline object, populated from template in the configuration directory."""
+        boto_objects, boto_parameters = _read_template(template_name, log_location)
+        return cls(contents=boto_objects, parameter_definitions=boto_parameters,
+                   name=name, description=description, tags=tags, param_values=param_values)
+
+
+class AsiaqDataPipelineManager(object):
+    "List, retrieve, store and delete pipelines."
+    def __init__(self, client=None):
+        self._dp_client = client or boto3.client("datapipeline")
+
+    def save(self, pipeline):
+        "Save or update the pipeline object in AWS."
+        if not pipeline.is_persisted():  # if this is a new pipeline, save the metadata first
+            unique_id = "pipeline_init_%s" % datetime.now().isoformat()  # make the create call idempotent
+            created = throttled_call(self._dp_client.create_pipeline,
+                                     name=pipeline._name, uniqueId=unique_id,
+                                     description=pipeline._description, tags=pipeline._tags or [])
+            pipeline._id = created['pipelineId']
+        # regardless, save the pipeline content:
+        resp = throttled_call(self._dp_client.put_pipeline_definition,
+                              pipelineId=pipeline._id,
+                              pipelineObjects=pipeline._objects or [],
+                              parameterObjects=pipeline._params or [],
+                              parameterValues=pipeline._param_values or [])
+        return resp
+
+    # possibly worth having but not currently needed: set_tags
+    # http://boto3.readthedocs.io/en/latest/reference/services/datapipeline.html#DataPipeline.Client.add_tags
+
+    def fetch(self, pipeline_id):
+        "Fetch a pipeline (metadata and content) from AWS by ID."
+        contents = throttled_call(self._dp_client.get_pipeline_definition, pipelineId=pipeline_id,
+                                  version='latest')
+        _LOG.debug("Contents for %s: %s", pipeline_id, contents)
+        meta_resp = throttled_call(self._dp_client.describe_pipelines, pipelineIds=[pipeline_id])
+        meta = meta_resp['pipelineDescriptionList'][0]
+        return AsiaqDataPipeline(
+            pipeline_id=pipeline_id, name=meta['name'], description=meta.get('description'),
+            tags=meta.get('tags'), metadata=meta.get('fields'),
+            contents=contents['pipelineObjects'],
+            parameter_definitions=contents.get('parameterObjects'),
+            param_values=contents.get('parameterValues')
+        )
+
+    def fetch_content(self, pipeline):
+        "Populate the pipeline content fields of this pipeline object with the information fetched from AWS."
+        if pipeline.has_content():
+            raise Exception("Content already fetched or locally generated for this pipeline object")
```

Comment:
Wouldn't object to it, just wasn't sure what I wanted it to be—we tend to have very specific exceptions and not re-use them in useful ways.  Any suggestions?  Or should I just create one and call it a day?